contributingGuides/TYPESCRIPT_STYLE.md

```diff
@@ -0,0 +1,897 @@
+
+- `index.native.ts` - Native (iOS and Android)
+- `index.android.ts` - Android
+- `index.ios.ts` - iOS
+- `index.website.ts` - Web
+- `index.desktop.ts` - Desktop
+
+Note that `index.ts` should contain the default implementation, and only platform-specific implementations should be done in their respective files. i.e: If you have mobile-specific implementation in `index.native.ts`, then the desktop/web implementation can be contained in a shared `index.ts`.
+
+For each platform-specific module, create shared type definitions in a separate `types.ts` file and place it in the same folder. `types.ts` has to export shared types which are **compatible with all platform-specific implementations**. Declare component props, return types, and other common types in this file.
+
+```ts
+type ComponentProps = {
+    foo: string;
+    bar: number;
+};
+
+type Helpers = {
+    doSomething: () => void;
+};
+
+export type { ComponentProps, Helpers };
+```
+
+In case there is no default implementation, you have to create a `index.d.ts` declaration file. When importing the module from other files, TypeScript will automatically pick up the type definitions from `index.d.ts`. Be careful when defining `index.d.ts` as declaration files aren't checked by the TypeScript compiler (with `skipLibCheck: true` - [source](https://www.typescriptlang.org/tsconfig#skipLibCheck)).
+
+Example with a default implementation:
+
+```ts
+// types.ts
+type VisibilityInterface = {
+    isVisible: () => boolean;
+    hasFocus: () => boolean;
+};
+
+export default VisibilityInterface;
+
+// index.ts
+function isVisible() {
+    return document.visibilityState === 'visible';
+}
+
+function hasFocus() {
+    return document.hasFocus();
+}
+
+const Visibility: VisibilityInterface = {
+    isVisible,
+    hasFocus,
+};
+
+export default Visibility;
+
+// index.native.ts
+function isVisible() {
+    return AppState.currentState === 'active';
+}
+
+function hasFocus() {
+    return true;
+}
+
+const Visibility: VisibilityInterface = {
+    isVisible,
+    hasFocus,
+};
+
+export default Visibility;
+```
+
+Example with no default implementation:
+
+```ts
+// types.ts
+type Platform = 'android' | 'ios' | 'web' | 'desktop';
+export default Platform;
+
+// index.d.ts
+export default function getPlatform(): Platform;
+
+// index.ios.ts
+export default function getPlatform(): Platform {
+  return "ios";
+}
+
+// index.android.ts
+export default function getPlatform(): Platform {
+  return "android";
+}
+
+// index.desktop.ts
+export default function getPlatform(): Platform {
+  return "desktop";
+}
+
+// index.website.ts
+export default function getPlatform(): Platform {
+  return "website";
+}
+```
+
+Refer to the React Native documentation for more information about [Platform Specific Code](https://reactnative.dev/docs/platform-specific-code).
+
+## Naming Conventions
```

Comment:
It seems like most of this might be possible to enforce via https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/naming-convention.md, maybe we can implement that in https://github.com/Expensify/App/pull/20179 and then remove this section. In general I don't think the guidelines would be needed if they were enforced by lint rules.

Reply:
I think this eslint rule is a nice addition to this section. Even with the rule I would leave this part in the guideline.

> It seems like most of this might be possible to enforce via [link](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/naming-convention.md)

I don't think it is possible to enforce CONSTANT_CASE to only work for global-level constant values. Other should be fine I believe 👍

Reply:
I never used this rule before. I'm not sure if we need it as I haven't seen any naming convention discrepancies in the codebase but it is nice addition and I can add it to the Typescript setup PR!

@hayata-suenaga @fabioh8010 thoughts?

Reply:
@roryabraham, great find! @blazejkustra do you think you can figure out options for this rule to make the rule conform  to our naming standard?

I think it's worth writing down even short sentences in the guideline even lint rules enforce some aspects of the guideline.

otherwise, we also don't need section for `Enum` or `any` or the usage of `@ts-ignore`

Reply:
@blazejkustra I think you last commit about the eslint rule solve this discussion as well, isn't?

Reply:
I tried and came up with this rule (added it to the Typescript setup PR):
```ts
'@typescript-eslint/naming-convention': [
  'error',
  {
      selector: 'variable',
      format: ['camelCase', 'UPPER_CASE', 'PascalCase'],
  },
  {
      selector: ['property', 'function'],
      format: ['camelCase', 'PascalCase'],
  },
  {
      selector: ['enumMember', 'typeLike', 'typeParameter'],
      format: ['PascalCase'],
  },
  {
      selector: ['parameter', 'method'],
      format: ['camelCase'],
  },
],
```

Explanations:
- `variable`: `camelCase` for basic variables, `UPPER_CASE` for global-level constants, `PascalCase` for React Components (we need it when using `forwardRef`)
- `property`, `function`: 'camelCase' for default functions and object properties. `PascalCase` for react components and Enum like object properties.
- `enumMember`, `typeLike`, `typeParameter`: `PascalCase` for all types.
- `parameter`, `method`: `camelCase` for class methods and function parameters

This rule doesn't fully conform with our naming standards: 
- Global-level constants can be written with `camelCase` or `PascalCase`.
- Object properties can be written in `PascalCase`.
- Functions (not components) can be written in `PascalCase`.

There are more inconsistencies like this due to the fact that some selectors are used in multiple contexts. It doesn't fully conform with our naming standards, it's still helpful and eliminates some of the mistakes that could happen.

Thoughts @fabioh8010 @hayata-suenaga @roryabraham?

Reply:
LGTM! In this case, we should keep the conventions in the guideline as it's not possible to cover all cases with the rule.

Reply:
We should put non-typescript naming conventions inside our [existing JavaScript style guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md).

I know linting of the naming convention was made possible because we introduced TS eslint plugin, but let's keep JavaScript stuff in `STYLE.md` instead of TypeScript guideline